include/oneapi/dpl/internal/distributed_ranges_impl/sp/algorithms/exclusive_scan.hpp

```diff
@@ -0,0 +1,239 @@
+#include "../../detail/onedpl_direct_iterator.hpp"
+#include "../allocators.hpp"
+#include "../detail.hpp"
+#include "../init.hpp"
+#include "../vector.hpp"
+#include "../views/views.hpp"
+#include "execution_policy.hpp"
+
+namespace oneapi::dpl::experimental::dr::sp
+{
+
+template <typename ExecutionPolicy, distributed_contiguous_range R, distributed_contiguous_range O, typename U,
+          typename BinaryOp>
+void
+exclusive_scan_impl_(ExecutionPolicy&& policy, R&& r, O&& o, U init, BinaryOp binary_op)
+{
+    using T = stdrng::range_value_t<O>;
+
+    static_assert(std::is_same_v<std::remove_cvref_t<ExecutionPolicy>, sycl_device_collection>);
+
+    auto zipped_view = views::zip(r, o);
+    auto zipped_segments = zipped_view.zipped_segments();
+
+    U* d_inits = sycl::malloc_device<U>(stdrng::size(zipped_segments), devices()[0], context());
+
+    std::vector<sycl::event> events;
+
+    std::size_t segment_id = 0;
+    for (auto&& segs : zipped_segments)
+    {
+        auto&& [in_segment, out_segment] = segs;
+
+        auto last_element = stdrng::prev(stdrng::end(__detail::local(in_segment)));
+        auto dest = d_inits + segment_id;
+
+        auto&& q = __detail::queue(ranges::rank(in_segment));
+
+        auto e = q.single_task([=] { *dest = *last_element; });
+        events.push_back(e);
+        segment_id++;
+    }
+
+    __detail::wait(events);
+    events.clear();
+
+    std::vector<U> inits(stdrng::size(zipped_segments));
+
+    copy(d_inits, d_inits + inits.size(), inits.data() + 1);
+
+    sycl::free(d_inits, context());
+
+    // when empty range fails due to: https://github.com/oneapi-src/distributed-ranges/issues/791
+    inits[0] = init;
+
+    auto root = devices()[0];
+    device_allocator<T> allocator(context(), root);
+    vector<T, device_allocator<T>> partial_sums(std::size_t(zipped_segments.size()), allocator);
+
+    // DPL futures must be kept alive, since in the future their destruction
+    // may trigger a synchronization.
+    using dpl_future_type = decltype(exclusive_scan_async(
+        __detail::dpl_policy(0), dr::__detail::direct_iterator(stdrng::begin(std::get<0>(*zipped_segments.begin()))),
+        dr::__detail::direct_iterator(stdrng::begin(std::get<0>(*zipped_segments.begin()))),
+        dr::__detail::direct_iterator(stdrng::begin(std::get<1>(*zipped_segments.begin()))), init, binary_op));
+
+    std::vector<dpl_future_type> futures;
+
+    segment_id = 0;
+    for (auto&& segs : zipped_segments)
+    {
+        auto&& [in_segment, out_segment] = segs;
+
+        auto&& q = __detail::queue(ranges::rank(in_segment));
+        auto&& local_policy = __detail::dpl_policy(ranges::rank(in_segment));
+
+        auto dist = stdrng::distance(in_segment);
+        assert(dist > 0);
+
+        auto first = stdrng::begin(in_segment);
+        auto last = stdrng::end(in_segment);
+        auto d_first = stdrng::begin(out_segment);
+
+        auto init = inits[segment_id];
```

Comment:
This `init` shadows the user-provided init value. I believe there needs to be explicit init handling for the first segment, similar to the `inclusive_scan`.

Reply:
I think this is correct as written?  On lines 49-75, we set `inits[i]` to be the last element of segment `i-1` (where `i > 0`).  Then on line 78 we set `inits[0]` to be the user-provided `init` value.

In the `exclusive_scan`, we "lose" the last element of each segment, and it serves as the `init` for the scan of the next segment.

We are definitely shadowing the global `init` parameter, which we could perhaps fix, but I believe the algorithm is correct, and we shouldn't need a local "no-init" scan like `inclusive_scan`.  (The global init or the "lost" final element from the last segment provide the init.)

Reply:
Ah, I suspected I was missing something. I think they shadowing of the `init` parameter just threw me off. Thanks for the clarification.